### ISSUE TITLE
Bug: no-restricted-imports Unable to check default import with importNamePattern

### ISSUE THREAD

**Issue by nikolay-borzov** [eslint/eslint#17861]
### Environment

Node version: 20.6.1
npm version: 9.8.1
Local ESLint version: 8.55.0
Global ESLint version: N/A
Operating System: Windows 10

### What parser are you using?

@typescript-eslint/parser

### What did you do?

<details>
<summary>Configuration</summary>

```
"no-restricted-imports": ["error", {
  "patterns": [{
    "group": ["@ds/icons/*"],
    // Require name ending with Icon
    "importNamePattern": "^(?!.*Icon$).*$",
  }]
}]
```
</details>

```js
import ThunderboltIcon from '@ds/icons/thunderbolt.svg';
import Thunderbolt from '@ds/icons/thunderbolt.svg';
import { ThunderboltNamed }  from '@ds/icons/thunderbolt.svg';
import { ThunderboltNamedIcon }  from '@ds/icons/thunderbolt.svg';
```

### What did you expect to happen?

Expected default import name (`ThunderboltIcon`, `Thunderbolt`) to be tested with `importNamePattern`

### What actually happened?

`"default"` is tested with `importNamePattern`

> 'default' import from '@ds/icons/thunderbolt.svg' is restricted from being used by a pattern.

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IFwibm8tcmVzdHJpY3RlZC1pbXBvcnRzXCI6IFtcImVycm9yXCIsIHtcbiAgICBcInBhdHRlcm5zXCI6IFt7XG4gICAgICBcImdyb3VwXCI6IFtcIkBkcy9pY29ucy8qXCJdLFxuICAgICAgXCJpbXBvcnROYW1lUGF0dGVyblwiOiBcIl4oPyEuKkljb24kKS4qJFwiLFxuICAgIH1dXG59XVxuKi9cblxuaW1wb3J0IFRodW5kZXJib2x0SWNvbiBmcm9tICdAZHMvaWNvbnMvdGh1bmRlcmJvbHQuc3ZnJztcbmltcG9ydCBUaHVuZGVyYm9sdCBmcm9tICdAZHMvaWNvbnMvdGh1bmRlcmJvbHQuc3ZnJztcbmltcG9ydCB7IFRodW5kZXJib2x0TmFtZWQgfSAgZnJvbSAnQGRzL2ljb25zL3RodW5kZXJib2x0LnN2Zyc7XG5pbXBvcnQgeyBUaHVuZGVyYm9sdE5hbWVkSWNvbiB9ICBmcm9tICdAZHMvaWNvbnMvdGh1bmRlcmJvbHQuc3ZnJzsiLCJvcHRpb25zIjp7ImVudiI6e30sInJ1bGVzIjp7fSwicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9LCJlY21hVmVyc2lvbiI6MTIsInNvdXJjZVR5cGUiOiJtb2R1bGUifX19

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

If it isn't possible to get actual name of the `default` import then
- Testing "default" should be ignored
or
- An error about misusing config is reported

**Comment by mdjermanovic** [comment]
Hi @nikolay-borzov, thanks for the issue!

I think this works as intended. `importNamePattern` and `importNames` options relate to names exported from the module, not to local names in modules that import it.

In the same way, `"importNames": ["default"]` disallows default imports:

```js
/* eslint "no-restricted-imports": ["error", {
    "patterns": [{
      "group": ["@ds/icons/*"],
      "importNames": ["default"],
    }]
}]
*/

import ThunderboltIcon from '@ds/icons/thunderbolt.svg'; // 'default' import from '@ds/icons/thunderbolt.svg' is restricted from being used by a pattern
import Thunderbolt from '@ds/icons/thunderbolt.svg'; // 'default' import from '@ds/icons/thunderbolt.svg' is restricted from being used by a pattern
import { ThunderboltNamed }  from '@ds/icons/thunderbolt.svg';
import { ThunderboltNamedIcon }  from '@ds/icons/thunderbolt.svg';
```

[Playground demo](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IFwibm8tcmVzdHJpY3RlZC1pbXBvcnRzXCI6IFtcImVycm9yXCIsIHtcbiAgICBcInBhdHRlcm5zXCI6IFt7XG4gICAgICBcImdyb3VwXCI6IFtcIkBkcy9pY29ucy8qXCJdLFxuICAgICAgXCJpbXBvcnROYW1lc1wiOiBbXCJkZWZhdWx0XCJdLFxuICAgIH1dXG59XVxuKi9cblxuaW1wb3J0IFRodW5kZXJib2x0SWNvbiBmcm9tICdAZHMvaWNvbnMvdGh1bmRlcmJvbHQuc3ZnJzsgLy8gJ2RlZmF1bHQnIGltcG9ydCBmcm9tICdAZHMvaWNvbnMvdGh1bmRlcmJvbHQuc3ZnJyBpcyByZXN0cmljdGVkIGZyb20gYmVpbmcgdXNlZCBieSBhIHBhdHRlcm5cbmltcG9ydCBUaHVuZGVyYm9sdCBmcm9tICdAZHMvaWNvbnMvdGh1bmRlcmJvbHQuc3ZnJzsgLy8gJ2RlZmF1bHQnIGltcG9ydCBmcm9tICdAZHMvaWNvbnMvdGh1bmRlcmJvbHQuc3ZnJyBpcyByZXN0cmljdGVkIGZyb20gYmVpbmcgdXNlZCBieSBhIHBhdHRlcm5cbmltcG9ydCB7IFRodW5kZXJib2x0TmFtZWQgfSAgZnJvbSAnQGRzL2ljb25zL3RodW5kZXJib2x0LnN2Zyc7XG5pbXBvcnQgeyBUaHVuZGVyYm9sdE5hbWVkSWNvbiB9ICBmcm9tICdAZHMvaWNvbnMvdGh1bmRlcmJvbHQuc3ZnJzsiLCJvcHRpb25zIjp7ImVudiI6e30sInJ1bGVzIjp7fSwicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9LCJlY21hVmVyc2lvbiI6MTIsInNvdXJjZVR5cGUiOiJtb2R1bGUifX19)

If you want to allow only default import and named imports that end with "Icon", you can update the pattern to `"^(?!.*Icon$|default$).*$"`:

```js
/* eslint "no-restricted-imports": ["error", {
    "patterns": [{
      "group": ["@ds/icons/*"],
      "importNamePattern": "^(?!.*Icon$|default$).*$",
    }]
}]
*/

import ThunderboltIcon from '@ds/icons/thunderbolt.svg'; // ok
import Thunderbolt from '@ds/icons/thunderbolt.svg'; // ok
import { ThunderboltNamed }  from '@ds/icons/thunderbolt.svg'; // 'ThunderboltNamed' import from '@ds/icons/thunderbolt.svg' is restricted from being used by a pattern.
import { ThunderboltNamedIcon }  from '@ds/icons/thunderbolt.svg'; // ok
```

[Playground demo](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IFwibm8tcmVzdHJpY3RlZC1pbXBvcnRzXCI6IFtcImVycm9yXCIsIHtcbiAgICBcInBhdHRlcm5zXCI6IFt7XG4gICAgICBcImdyb3VwXCI6IFtcIkBkcy9pY29ucy8qXCJdLFxuICAgICAgXCJpbXBvcnROYW1lUGF0dGVyblwiOiBcIl4oPyEuKkljb24kfGRlZmF1bHQkKS4qJFwiLFxuICAgIH1dXG59XVxuKi9cblxuaW1wb3J0IFRodW5kZXJib2x0SWNvbiBmcm9tICdAZHMvaWNvbnMvdGh1bmRlcmJvbHQuc3ZnJzsgLy8gb2tcbmltcG9ydCBUaHVuZGVyYm9sdCBmcm9tICdAZHMvaWNvbnMvdGh1bmRlcmJvbHQuc3ZnJzsgLy8gb2tcbmltcG9ydCB7IFRodW5kZXJib2x0TmFtZWQgfSAgZnJvbSAnQGRzL2ljb25zL3RodW5kZXJib2x0LnN2Zyc7IC8vICdUaHVuZGVyYm9sdE5hbWVkJyBpbXBvcnQgZnJvbSAnQGRzL2ljb25zL3RodW5kZXJib2x0LnN2ZycgaXMgcmVzdHJpY3RlZCBmcm9tIGJlaW5nIHVzZWQgYnkgYSBwYXR0ZXJuLlxuaW1wb3J0IHsgVGh1bmRlcmJvbHROYW1lZEljb24gfSAgZnJvbSAnQGRzL2ljb25zL3RodW5kZXJib2x0LnN2Zyc7IC8vIG9rIiwib3B0aW9ucyI6eyJlbnYiOnt9LCJydWxlcyI6e30sInBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fSwiZWNtYVZlcnNpb24iOjEyLCJzb3VyY2VUeXBlIjoibW9kdWxlIn19fQ==)

**Comment by nikolay-borzov** [comment]
Thanks for explanation!

My problem is that I want to restrict default import names. For me testing `importNames` and `importNamePattern` against "default" doesn't make much sense since "default" is not visible in code

**Comment by mdjermanovic** [comment]
> My problem is that I want to restrict default import names.

You would like to enforce naming for the default import?

```js
import Thunderbolt from '@ds/icons/thunderbolt.svg'; // not ok

import ThunderboltIcon from '@ds/icons/thunderbolt.svg'; // ok
```

**Comment by nikolay-borzov** [comment]
Right

**Comment by mdjermanovic** [comment]
Thanks for clarifying the expected behavior.

Restricting local names is not a responsibility of this rule. This rule restricts importing modules or certain exports from modules. It doesn't enforce naming conventions for local names. In the above example, `Thunderbolt` and `ThunderboltIcon` are local names. The exported name is `default`, and thus all configured restrictions are checked against that name. 

It works the same for named imports:

```js
/* eslint "no-restricted-imports": ["error", {
    "patterns": [{
      "group": ["@ds/icons/*"],
      "importNamePattern": "^foo",
    }]
}]
*/

import { foo as bar } from '@ds/icons/thunderbolt.svg'; // error, because this imports `foo' 
import { bar as foo } from '@ds/icons/thunderbolt.svg'; // ok, because this imports `bar`, and `foo` is just a local name

```
